### ISSUE TITLE
List examples where IDEA failes to parse valid Freestyle code

### ISSUE THREAD

**Issue by raulraja** [47deg/freestyle#401]
We need to compile a list of all instances in which the IDEA parser fails to properly highlight Freestyle code in the following format:
```
<IDEA error message>
- Valid compiling code.
- Screenshot.
```
A good start is opening Freestyle in IDEA and navigating through the sources and tests to find red marked blocks of code where inference is failing or false compilations errors are being reported.

**Comment by dominv** [comment]
- **Wrong number of type parameters. Expected: 0, actual: 1**
```scala 
  def bootstrap[F[_]](
      implicit app: App[F],
```

https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/http/TodoListApp.scala#L60

![2017-09-25-175623_433x63_scrot](https://user-images.githubusercontent.com/3943031/30818220-3d904a1e-a21b-11e7-95d6-1e627c78bd5b.png)

![2017-09-25-175635_306x76_scrot](https://user-images.githubusercontent.com/3943031/30818303-7c2d8d04-a21b-11e7-9a29-eb46df1f0272.png)

Macro Expanded: 

![2017-09-25-175847_994x227_scrot](https://user-images.githubusercontent.com/3943031/30818305-80278de2-a21b-11e7-8890-790618ecfd41.png)

**Comment by dominv** [comment]
- **Cannot resolve symbol interpret**
```scala
   val server = Await.result(bootstrap[App.Op].interpret[Future])
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/http/TodoListApp.scala#L85

![2017-09-25-180630_882x181_scrot](https://user-images.githubusercontent.com/3943031/30818601-4a02c258-a21c-11e7-8d98-60ec2555c8e1.png)

**Comment by dominv** [comment]
- **Wrong number of type parameters. Expected: 0, actual: 1**
```scala
class AppApi[F[_]](implicit service: AppServices[F], handler: F ~> Future) {
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/http/apis/AppApi.scala#L31

![2017-09-25-181024_610x44_scrot](https://user-images.githubusercontent.com/3943031/30818782-d207c8ba-a21c-11e7-8e86-820529e7a9df.png)

**Comment by dominv** [comment]
- **Type mismatch, expected: Input, actual: Free[({ type λ[β$0$] = FreeApplicative[FF$974, β$0$] })#λ, Output[Int]]**
```scala 
  val reset: Endpoint[Int] =
    post("reset") {
      service.reset.map(Ok(_))
    }
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/http/apis/AppApi.scala#L35

![2017-09-25-181742_353x88_scrot](https://user-images.githubusercontent.com/3943031/30819062-da6634be-a21d-11e7-9658-4cc7644a20f1.png)

**Comment by dominv** [comment]
- **Cannot resolve symbol query**
```scala
  def list: ConnectionIO[List[AppModel]] = {
    val data =
      sql"""SELECT lists.title, lists.tag_id, lists.id, tags.name, tags.id, items.item, items.todo_list_id, items.completed, items.id FROM todo_lists AS lists INNER JOIN tags ON lists.tag_id = tags.id LEFT JOIN todo_items AS items ON lists.id = items.todo_list_id"""
        .query[(TodoList, Tag, (Option[String], Option[Int], Option[Boolean], Option[Int]))]
        .list
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/runtime/handlers/H2AppRepositoryHandler.scala#L29

![2017-09-25-183025_953x110_scrot](https://user-images.githubusercontent.com/3943031/30819650-afb09708-a21f-11e7-8465-b51f3ce2a95a.png)

**Comment by dominv** [comment]
- **Cannot resolve symbol λ**
```scala
  implicit val connectionIO2Task: ConnectionIO ~> Task =
    λ[ConnectionIO ~> Task](_.transact(xa))
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/runtime/implicits.scala#L88

![2017-09-25-184704_516x59_scrot](https://user-images.githubusercontent.com/3943031/30820342-f12f4e0c-a221-11e7-9a68-7013bd11816c.png)

**Comment by dominv** [comment]
- **Cannot resolve symbol transact**
```scala
  implicit val connectionIO2Task: ConnectionIO ~> Task =
    λ[ConnectionIO ~> Task](_.transact(xa))
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/runtime/implicits.scala#L88

![2017-09-25-184704_516x59_scrot](https://user-images.githubusercontent.com/3943031/30820342-f12f4e0c-a221-11e7-9a68-7013bd11816c.png)

**Comment by dominv** [comment]
- **Expression of type FunctionK[TodoItemRepository.Op, Future] doesn't conform to expected type cats.~>[TodoItemRepository.Op, Future]**
```scala
  implicit val todoItemRepositoryHandler: TodoItemRepository.Op ~> Future =
    new H2TodoItemRepositoryHandler andThen connectionIO2Task andThen task2Future
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/runtime/implicits.scala#L94

![2017-09-25-185054_647x56_scrot](https://user-images.githubusercontent.com/3943031/30820545-7d93e4b6-a222-11e7-9a52-2b8e8a6cac57.png)

**Comment by dominv** [comment]
- **Type ErrorM takes type parameters**
```scala
@module
trait AppServices {

  val repo: AppRepository
  val tagServices: TagService
  val todoItemServices: TodoItemService
  val todoListServices: TodoListService
  val error: ErrorM
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/services/AppServices.scala#L33

![2017-09-25-185618_360x150_scrot](https://user-images.githubusercontent.com/3943031/30820771-3b70b392-a223-11e7-9e5c-65ccdc34c3fe.png)

Macro Expanded:

![2017-09-25-185805_1113x187_scrot](https://user-images.githubusercontent.com/3943031/30820863-889b7f30-a223-11e7-994c-43935d6bf662.png)

**Comment by dominv** [comment]
- **Expression of type FreeApplicative.FA[FF$510, List[TodoForm]] doesn't conform to expected type AppServices.this.FS.Seq[List[TodoForm]]**
```scala
  val list: FS.Seq[List[TodoForm]] = repo.list.map(_.groupBy(x => (x._1, x._2)).map {
    case ((todoList, tag), list) =>
      TodoForm(todoList, tag, list.flatMap(_._3))
  }.toList)
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/services/AppServices.scala#L42

![2017-09-25-190158_686x87_scrot](https://user-images.githubusercontent.com/3943031/30821002-077cd092-a224-11e7-8475-d57ce6fb1944.png)

**Comment by dominv** [comment]
- **Cannot resolve symbol getOrElse**
```scala
  def insert(form: TodoForm): FS.Seq[TodoForm] =
    for {
      tag <- tagServices.insert(form.tag)
      t   <- error.either[Tag](tag.toRight(new NoSuchElementException("Could not create Tag")))

      list <- todoListServices.insert(form.list.copy(tagId = t.id))
      l <- error.either[TodoList](
        list.toRight(new NoSuchElementException("Could not create TodoList")))

      i <- todoItemServices.batchedInsert(form.items.map(_.copy(todoListId = l.id)))
      items = i.sequence
    } yield form.copy(list = l, tag = t, items = items getOrElse form.items)
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/services/AppServices.scala#L58

![2017-09-25-190547_758x205_scrot](https://user-images.githubusercontent.com/3943031/30821130-8e51c758-a224-11e7-9405-a4c4e06db69f.png)

**Comment by dominv** [comment]
- **Cannot resolve symbol map**
```scala
def destroy(form: TodoForm): FS.Seq[Int] = {
    val todoItemIds: Option[List[Int]] = form.items.map(_.id).sequence
    val todoListId: Option[Int]        = form.list.id
    val todoTagId: Option[Int]         = form.tag.id

    val program = for {
      a <- todoItemIds.map(todoItemServices.batchedDestroy)
      b <- todoListId.map(todoItemServices.destroy)
      c <- todoTagId.map(tagServices.destroy)
    } yield List(a, b, c)

    program.map { x =>
      x.sequenceU
        .map(_.sum)
    } getOrElse (throw new NoSuchElementException("Could not delete"))
  }
```
https://github.com/frees-io/freestyle-examples/blob/73573e80ef1d7894de33fc33ccbbce2182e70546/todolist/src/main/scala/todo/services/AppServices.scala#L86

![2017-09-25-190807_606x283_scrot](https://user-images.githubusercontent.com/3943031/30821239-e4fc82be-a224-11e7-924f-cee7e1976103.png)

**Comment by anamariamv** [comment]
[Issue in youtrack](https://youtrack.jetbrains.com/issue/SCL-12582)
